src/change_trust_asset.js

```diff
@@ -0,0 +1,280 @@
+import xdr from './generated/stellar-xdr_generated';
+import { Asset } from './asset';
+import { Keypair } from './keypair';
+import { LiquidityPoolFeeV18, getLiquidityPoolId } from './liquidity_pool_id';
+import { StrKey } from './strkey';
+
+/**
+ * ChangeTrustAsset class represents a trustline change to either a liquidity
+ * pool, the native asset (`XLM`) or an issued asset with an asset code / issuer
+ * account ID pair.
+ *
+ * In case of the native asset, the code will represent `XLM` while the issuer
+ * and liquidityPoolParameters will be empty. For an issued asset, the code and
+ * issuer will be valid and liquidityPoolParameters will be empty. For liquidity
+ * pools, the liquidityPoolParameters will be valid and the remaining two fields
+ * will be empty.
+ *
+ * @constructor
+ * @param {string} code - The asset code.
+ * @param {string} issuer - The account ID of the asset issuer.
+ * @param {LiquidityPoolParameters} liquidityPoolParameters – The liquidity pool parameters.
+ * @param {Asset} liquidityPoolParameters.assetA – The first asset in the Pool, it must respect the rule assetA < assetB. See `Asset.compare(A, B)` for more details on how assets are sorted.
+ * @param {Asset} liquidityPoolParameters.assetB – The second asset in the Pool, it must respect the rule assetA < assetB. See `Asset.compare(A, B)` for more details on how assets are sorted.
+ * @param {number} liquidityPoolParameters.fee – The liquidity pool fee. For now the only fee supported is `30`.
+ */
+export class ChangeTrustAsset {
+  constructor(code, issuer, liquidityPoolParameters) {
+    if (!code && !issuer && !liquidityPoolParameters) {
+      throw new Error(
+        'Must provide either code, issuer or liquidityPoolParameters'
+      );
+    }
+
+    if ((code || issuer) && liquidityPoolParameters) {
+      throw new Error(
+        'Must provide either code (and optionally issuer) or liquidityPoolParameters but not both'
+      );
+    }
+
+    // Validate code & issuer.
+    if ((code || issuer) && !/^[a-zA-Z0-9]{1,12}$/.test(code)) {
+      throw new Error(
+        'Asset code is invalid (maximum alphanumeric, 12 characters at max)'
+      );
+    }
+    if (code && String(code).toLowerCase() !== 'xlm' && !issuer) {
+      throw new Error('Issuer cannot be null');
+    }
+    if (issuer && !StrKey.isValidEd25519PublicKey(issuer)) {
+      throw new Error('Issuer is invalid');
+    }
+
+    this.code = code;
+    this.issuer = issuer;
+    if (code) {
+      // return early if it's an asset with code.
+      return;
+    }
+
+    // Validate liquidity pool params.
+    const { assetA, assetB, fee } = liquidityPoolParameters;
+    if (!assetA || !(assetA instanceof Asset)) {
+      throw new Error('assetA is invalid');
+    }
+    if (!assetB || !(assetB instanceof Asset)) {
+      throw new Error('assetB is invalid');
+    }
+    if (Asset.compare(assetA, assetB) !== -1) {
+      throw new Error('Assets are not in lexicographic order');
+    }
```

Comment:
Out of scope but worth asking: Should we enforce this, or would it be better to just be helpful and order them ourselves?

Reply:
In [this offline thread](https://stellarfoundation.slack.com/archives/C01PNKPSY3H/p1629393624105600) we decided not to change the order of the assets when building liquidity pool deposits, withdrawals or when building the liquidity pool ID. Since the logic here is pretty similar to the getLiquidityPoolId method I kept the logic consistent for both of them.

The argument was that changing the assets orders at any point could bring unforeseen downsides. For instance, if you build a transaction with: `[LP ChangeTrust, LP Deposit, LP Withdrawal]` and you use the same assets for all these operations you should not change the asset order for any of them at the risk of failing the transaction or messing with the deposit/withdrawal amounts.